plrustc/plrustc/src/main.rs

```diff
@@ -127,124 +70,101 @@ fn main() {
-) -> ! {
-    let mut extra_args = vec![];
-    if target_crate {
-        extra_args.push("--cfg=plrust");
+/// Get the sysroot, looking from most specific to this invocation to the
+/// least.
+///
+/// - command line `--sysroot` arg (happens in caller)
+///
+/// - runtime environment
+///    - PLRUSTC_SYSROOT
+///    - SYSROOT
+///    - RUSTUP_HOME, RUSTUP_TOOLCHAIN
+///
+/// - sysroot from rustc in the path
+///
+/// - compile-time environment
+///    - PLRUSTC_SYSROOT
+///    - SYSROOT
+///    - RUSTUP_HOME, RUSTUP_TOOLCHAIN
+fn sysroot() -> Option<PathBuf> {
+    fn rustup_sysroot<H: ?Sized + AsRef<OsStr>, T: ?Sized + AsRef<Path>>(
+        home: &H,
+        toolchain: &T,
+    ) -> PathBuf {
+        let mut path = PathBuf::from(home);
+        path.push("toolchains");
+        path.push(toolchain);
+        path
     }
-    args.splice(1..1, extra_args.iter().map(ToString::to_string));
+    fn runtime_rustup_sysroot() -> Option<PathBuf> {
+        let home = std::env::var_os("RUSTUP_HOME")?;
+        let toolchain = std::env::var_os("RUSTUP_TOOLCHAIN")?;
+        Some(rustup_sysroot(&home, &toolchain))
+    }
+    fn compiletime_rustup_sysroot() -> Option<PathBuf> {
+        let home: &str = option_env!("RUSTUP_HOME")?;
+        let toolchain: &str = option_env!("RUSTUP_TOOLCHAIN")?;
+        Some(rustup_sysroot(&home, &toolchain))
+    }
+    fn rustc_on_path_sysroot() -> Option<PathBuf> {
+        std::process::Command::new("rustc")
+            .arg("--print=sysroot")
+            .output()
+            .ok()
+            .and_then(|out| String::from_utf8(out.stdout).ok())
+            .map(|s| PathBuf::from(s.trim()))
+    }
+    fn runtime_explicit_env() -> Option<PathBuf> {
+        let sysroot =
+            std::env::var_os("PLRUSTC_SYSROOT").or_else(|| std::env::var_os("SYSROOT"))?;
+        Some(PathBuf::from(sysroot))
+    }
+    fn compiletime_explicit_env() -> Option<PathBuf> {
+        let plrustc_sysroot: Option<&str> = option_env!("PLRUSTC_SYSROOT");
+        if let Some(plrustc_sysroot) = plrustc_sysroot {
+            return Some(plrustc_sysroot.into());
+        }
+        let sysroot: Option<&str> = option_env!("SYSROOT");
+        if let Some(sysroot) = sysroot {
+            return Some(sysroot.into());
+        }
+        None
+    }
+    runtime_explicit_env()
+        .or_else(runtime_rustup_sysroot)
+        .or_else(rustc_on_path_sysroot)
+        .or_else(compiletime_explicit_env)
+        .or_else(compiletime_rustup_sysroot)
+}
 
-    // Invoke compiler, and handle return code.
-    let exit_code = rustc_driver::catch_with_exit_code(move || {
+fn run_compiler<CB: Callbacks + Send>(mut args: Vec<String>, callbacks: &mut CB) -> ! {
+    args.splice(1..1, std::iter::once("--cfg=plrustc".to_string()));
+    std::process::exit(rustc_driver::catch_with_exit_code(move || {
         rustc_driver::RunCompiler::new(&args, callbacks).run()
-    });
-    std::process::exit(exit_code)
+    }));
 }
```

Comment:
Nice. -80 lines.

Reply:
Also a lot simpler conceptually. It no longer invokes itself!